src/main/java/sparksoniq/spark/iterator/flowr/ReturnClauseSparkIterator.java

```diff
@@ -49,8 +60,53 @@ public ReturnClauseSparkIterator(RuntimeTupleIterator child, RuntimeIterator exp
         return itemRDD;
     }
 
-    private JavaRDD<Item> itemRDD;
-    
-    private RuntimeTupleIterator _child;
+    @Override
+    protected boolean hasNextLocal() {
+        return _hasNext;
+    }
+
+    @Override
+    protected Item nextLocal() {
+        if(_hasNext == true){
+            Item result = _nextLocalResult;  // save the result to be returned
+            setNextLocalResult();            // calculate and store the next result
+            return result;
+        }
+        throw new IteratorFlowException("Invalid next() call in Object Lookup", getMetadata());
+    }
+
+    @Override
+    protected void openLocal(DynamicContext context) {
+        this._currentDynamicContext = context;
+
+        _iterator.open(context);
+        setNextLocalResult();
+    }
+
+    private void setNextLocalResult() {
+        if (_iterator.hasNext()) {
```

Comment:
This is actually slight more complex (but still relatively simple).

You need to:
1 iterate on the _child iterator (the clause above), getting Tuples
2 for each Tuple obtained above, chain the _currentDynamicContext with the dynamic context obtained from the Tuple (something like that is done in the filter expression implementation, for example).
3 open the _iterator iterator, passing the above chained dynamic context
4 iterate on the _iterator iterator until hasNext() is false, then go back to 2 and continue.

Note that nothing needs to be materialized and this is just a bunch of if-then-elses:

For every call to setNextLocalResult(), you simply check whether _iterator is initialized and has more items.

If it does, you just call next() and assign it to _nextLocalResult.

If it doesn't, then you check whether_child has more tuples (if it is not initialized, you do so).

If it does, you call next(), get the Tuple, initialize a new dynamic context, reset _iterator with that context and assign the first item to _nextLocalResult.

If _child has no more tuples, then this is the end and _hasNext becomes false.

Reply:
Thanks a lot for the detailed feedback, it was very helpful. I have just pushed an implementation which I believe covers most of the points of execution you have listed. I'm not sure if I have got the dynamic context chaining right, however.